[flang][hlfir] address char_convert issues as mentioned in #64315

cabreraam opened the pull request:
As shown in #64315, there were issues when trying to lower two flavors of Fortran code that implicitly perform `CHARACTER` conversion. Specifically, I address the following issues:

```
error: loc("repro.f90":9:3): 'fir.char_convert' op operand #0 must be any reference, but got '!fir.boxchar<4>'
error: loc("repro.f90":15:3): 'fir.char_convert' op operand #0 must be any reference, but got '!hlfir.expr<!fir.char<1>>'
```

@vzakhari @jeanPerier I borrowed a helper function from [HLFIRTools.cpp](https://github.com/llvm/llvm-project/blob/e04bf911113566cec2e399f0909aed4e5dc309a0/flang/lib/Optimizer/Builder/HLFIRTools.cpp#L168-L183). The function is `static` so I'm unable to link against it in `flang/lib/Lower/ConvertExprToHLFIR.cpp`. Is there a cleaner way that what I've done, which is just copy/paste code?



Thanks,
Anthony

jeanPerier commented on flang/lib/Lower/ConvertExprToHLFIR.cpp:
```diff
@@ -1435,8 +1450,15 @@ struct UnaryOp<
-      builder.create<fir::CharConvertOp>(loc, lhs.getFirBase(), origBufferSize,
-                                         dest);
+      auto src = lhs.getFirBase();
+      if (lhs.getType().isa<hlfir::ExprType>()) {
+        if (auto asExpr = lhs.getDefiningOp<hlfir::AsExprOp>()) {
+          src = asExpr.getVar();
+        }
```
You should not expect that AsExprOp this is the only way you will get an hlfir.expr coming in, this could have been the result of a concat....
You should use hlfir.associate instead to get a value in memory here which is what fir.char_convert expect.

In fact, since fir.char_convert is at the FIR level and not HLFIR, I think you should likely simply use 
[ hlfir::convertToAddress](https://github.com/llvm/llvm-project/blob/55accc82bec48acae769b086ad9a5dc29da77f02/flang/include/flang/Optimizer/Builder/HLFIRTools.h#L402) with lhs.getFortranElementType() as the target type.
You can call the returned cleanup, if any, after fir.char_convert.

This should cover all cases, remove the need from manually unboxing here (convertToAddress) does it, and remove the need from copy/pasting code.

cabreraam replied:
That suggestion did the trick! What do you think?

jeanPerier reviewed (changes requested):
Thanks for working on a fix!

jeanPerier commented on flang/lib/Lower/ConvertExprToHLFIR.cpp:
```diff
@@ -1435,8 +1434,10 @@ struct UnaryOp<
+      auto src = hlfir::convertToAddress(loc, builder, lhs,
+                                         lhs.getFortranElementType());
+      builder.create<fir::CharConvertOp>(loc, src.first.getCharBox()->getAddr(),
+                                         origBufferSize, dest);
```
Please add a call to src.second here if any to insert hlfir.destroy/hlfir.end_associate if some temporary buffer was created for the source. Otherwise looks good.

cabreraam replied:
Added the call. What do you think?

jeanPerier commented on flang/lib/Lower/ConvertExprToHLFIR.cpp:
```diff
@@ -1435,8 +1434,12 @@ struct UnaryOp<
+      builder.create<fir::CharConvertOp>(loc, src.first.getCharBox()->getAddr(),
+                                         origBufferSize, dest);
+      if (src.second.has_value())
+        src.second.value();
```
Are you sure this is calling the cleanup? I would actually expect `src.second.value()()` or `(*src.second)()` here.

Can you ensure there is an hlfir.end_associate created after the fir.char_convert for the following example:

```
subroutine charconvert3(c, c4)
  character(kind=1, len=*) :: c
  character(kind=4, len=*) :: c4
  c4 = c // c
end subroutine
```

cabreraam replied:
Yep, you are correct. I didn't actually call the `hlfir::cleanupFunction`.  I added another set of parens so now it's 
```cpp
src.second.value()()
```

I added your concat test and verified that `hlfir.end_associate` is called

jeanPerier reviewed (approved):
Thanks for addressing all the comments, LGTM

cabreraam commented:
> Thanks for addressing all the comments, LGTM

I messed up a push to the feature branch so I need to fix some stuff. I will make sure to fix this before closing and merging!!!

cabreraam committed:
[flang][hlfir] address char_convert issues as
mentioned in #64315
